### DISCUSSION TITLE
Trying to use USWDS tables in my Angular project, functionality missing unless table is visible on page load.

### DISCUSSION THREAD

**Discussion by dajones537** [uswds/uswds#5365] [Q&A] [answered]
I include USWDS as a package dependency in my package.json. In my angular.json, I include something like  this:

    "scripts": [
              "node_modules/uswds/dist/js/uswds.min.js"
    ]

For the uswds sortable table on my home page, everything looks great: the uswds.min.js created the appropriate sort buttons and such on my table. However, I have other pages / components thata are not initially visible. When I navigate to them, they are not given the sort headers buttons and such because the script has already run one time. If I manually copy the contents of uswds.min.js and paste and run them in chrome's console, the headers appear. 

What is the best way to have the table functionality script(s) ran whenever a new table that is made visible sometime after initial page load and thus the uswds.min.js script has already ran? This in the readme sounds relevant:

    import USWDS from "@uswds/uswds/js";
    const { characterCount, accordion } = USWDS; // deconstruct your components here

But I haven't been able to make any import of this sort work. For example, "import USWDS from "uswds/src/js/components";" results in this error: "Could not find a declaration for module [...]"

EDIT: I'm using uswds 2.14.0.

EDIT TWO:

I've gotten past this issue with the help below, thanks. Here is roughly what I look like now:

Define types in types.d.ts

    declare module 'uswds/src/js/components/table';
    declare module 'uswds/src/js/components/accordion';

Import inside my component

    import table from 'uswds/src/js/components/table';

Use it

    const myTable = document.getElementById(this.tableId);
    table.on(myTable);

I do have a new issue now, however, see below, in which clicking a sort button on a table that's nested inside another table results in toggleSort executing twice. Haven't solved that issue, but anyway I've gotten past the initial issue that this thread was about.

**Comment by jbouder** [comment]
Its been a little while since I've used USWDS in Angular, but I have been using in React, and in order to get it to work properly in react, I have to call the table.on() and table.off() functions appropriately when the table component loads/unloads to enable sorting functionality. There is an example here (scroll down to the Angular section): https://designsystem.digital.gov/documentation/developers/#js-customization-2.

This sort of thing also needs to be done on other more dynamic components like Accordion, Tooltip, etc.

**Reply by dajones537** [reply]
The Angular example in that link looks like the example in the readme:

`import USWDS from "@uswds/uswds/js"; const { characterCount, accordion } = USWDS; // deconstruct your components here`

When I try to import I get "Could not find a declaration for module [...]" errors. Without working imports, I don't think I can use the on or off functions yet. So for starters I need to figure out how to import, I think. Angular uses Typescript, and it seems like typescript doesn't know what to do with the plain old .js files inside uswds.

Edit: another example:

`Error: ../../core/components/dan/dan.component.ts:9:23 - error TS2307: Cannot find module '@uswds/uswds/js/usa-accordion' or its corresponding type declarations.`

**Reply by jbouder** [reply, marked as answer]
You might need to define types like I've done here: https://github.com/MetroStar/comet/blob/main/packages/comet-uswds/src/uswds/types.d.ts

**Reply by dajones537** [reply]
That sounds promising. I'll get back into this when I'm back at work this coming Tuesday, thanks.

**Reply by dajones537** [reply]
I've added a types.d.ts file with this in it:

`declare module '@uswds/uswds/packages/usa-table';`

Then I do this import in my component:

`import { table } from "@uswds/uswds/packages/usa-table";`

And I get the following error on compile:

`./src/app/core/components/fake/fake.component.ts:5:0-56 - Error: Module not found: Error: Package path ./packages/usa-table is not exported from package C:\Users\fake\Documents\GitHub\fakeParent\fakeUi\node_modules\@uswds\uswds (see exports field in C:\Users\fake\Documents\GitHub\fakeParent\fakeUi\node_modules\@uswds\uswds\package.json)`

EDIT: I may have had the wrong path in my module declaration. I used "packages" there because that's what I saw in folder structure, but per the exports config in uswds's package.json, I think I needed "js" like you had. Just compiled successfully with  this, like what you had:

`declare module '@uswds/uswds/js/usa-table';`

**Reply by dajones537** [reply]
Ok, so here's my current import:

`import { table } from "@uswds/uswds/js/usa-table";`

`
	aFunction(): void {
		console.log('We\'re now going to try to uswds-ify this table...', table);
		let ref = document.getElementById('aTable');
		table.on(ref);
		console.log('done')
	}
` 

In the console.log table shows as undefined, and I get an undefined error on table.on. However, if I use a breakpoint and mouseover table, table is NOT undefined.

**Reply by dajones537** [reply]
I changed my import from:

`import { table } from "@uswds/uswds/js/usa-table";`

to

`import table from "@uswds/uswds/js/usa-table";`

table is no longer undefined, and doing table.on(ref) gives this:

![image](https://github.com/uswds/uswds/assets/123194829/09dd45c6-79a8-40d2-8ca2-54d40a397086)

However, the sort buttons only work once:

![image](https://github.com/uswds/uswds/assets/123194829/da4743b2-b9eb-4b98-8afa-d0c1b424222b)

At this point clicking sort buttons doesn't do anything.

**Reply by dajones537** [reply]
Ok, I updated this in my angular.json

`
            "scripts": [
              "node_modules/uswds/dist/js/uswds.min.js",
              "src/assets/js/uswds/uswds-init.min.js",
              "src/assets/js/log-out.js"
        	]
`
to
`
            "scripts": [
              "src/assets/js/log-out.js"
        	]
`
Now everything looks fantastic, and seems to be working. I think I may be in good shape now. I'll return here later today and give a writeup of what my current state looks like to help anyone else who needs help, and mark you as answer. Thanks.

Edit: also how the heck do you format code in github? The <> button doesn't preserve linebreeaks.

**Reply by jbouder** [reply]
Awesome! Sounds like you did all the heavy lifting, but glad to help!

**Reply by dajones537** [reply]
Alright, so I'll writeup the stuff later, but just now I have another issue: I have a case where I need to show a nested table. I've found that whenever you click on a sortable header of the table that is inside the outer table, the toggleSort function is called twice, which has the effect of letting you only sort in one direction:
![image](https://github.com/uswds/uswds/assets/123194829/29ad8864-abf9-4971-827a-a354ac5964e0)
Whenever I show a given table I call table.on(ref) where ref is the table element that's being made visible. If, on the other hand, I were to show all tables, and then call table.on() without passing an argument, in that case all tables' sort headers work correctly.

Wondering if you or anyone has had a case where you had a nested table.

**Reply by jbouder** [reply]
Not something I've dealt with before and not sure if it would be an issue with the USWDS table or if its maybe an Angular specific thing. I would think if your tables have unique IDs and you're targeting each with its own specific 'ref' they shouldn't interfere with each other. That said though, if that doesn't help, the only other thought I have is...does the inner table need sorting? Maybe you can just get rid of it.

**Reply by dajones537** [reply]
My tables do have unique IDs. I could talk with folks about whether or not sorting is needed in the nested, but for now let's assume I'd like it to work. Here's a gif showing how the nested table works fine if you don't initialize the outer table, but when you initialize the outer and then the inner, the inner's sort always fires twice.
![table_trouble](https://github.com/uswds/uswds/assets/123194829/c85c7130-286b-4c9f-a555-44f375b56a06)

**Reply by jbouder** [reply]
Could be a bug. Might need to submit an issue and see if it can be verified by the USWDS team.

**Comment by ryparker** [comment]
The nested-table follow-up comes from overlapping initialization. I reproduced it with USWDS 3.14.0 without Angular: initializing the outer table also initializes headers inside the nested table. Initializing the inner table again adds another sort button and another delegated click listener. A click then bubbles through both listeners and sorts twice.

Initialize once on a shared container after its tables have rendered:

```js
import table from "@uswds/uswds/js/usa-table";

const root = document.querySelector("#tables");
table.on(root);

// When this container is removed:
// table.off(root);
```

With one initialization, the same reproduction sorts ascending on the first click and descending on the second. Showing an already-initialized table again does not need another `.on()` call. Also avoid loading the auto-initializing bundle alongside manual initialization of these tables.

One lifecycle detail: `.off(root)` removes listeners; it does not remove generated sort buttons. Repeated `.off()`/`.on()` calls on retained table markup are not a reset. The cleanup above assumes the container is being removed and a later mount gets fresh markup.
